beacon-chain/core/epoch/epoch_processing_test.go

```diff
@@ -1079,15 +1106,98 @@ func TestAttestationDelta_SomeAttested(t *testing.T) {
 		t.Fatal(err)
 	}
 
+	attestedBalance, err := AttestingBalance(state, atts)
+	totalBalance, _ := helpers.TotalActiveBalance(state)
+	if err != nil {
+		t.Fatal(err)
+	}
+
 	attestedIndices := []uint64{5, 754, 797, 1637, 1770, 1862, 1192}
+	for _, i := range attestedIndices {
+		base, err := baseReward(state, i)
+		if err != nil {
+			t.Errorf("Could not get base reward: %v", err)
+		}
+		// Base rewards for getting source right
+		wanted := 3 * (base * attestedBalance / totalBalance)
+		// Base rewards for proposer and attesters working together getting attestation
+		// on chain in the fatest manner
+		proposerReward := base / params.BeaconConfig().ProposerRewardQuotient
+		wanted += (base - proposerReward) * params.BeaconConfig().MinAttestationInclusionDelay
+		if rewards[i] != wanted {
+			t.Errorf("Wanted reward balance %d, got %d", wanted, rewards[i])
+		}
+		// Since all these validators attested, they shouldn't get penalized.
+		if penalties[i] != 0 {
+			t.Errorf("Wanted penalty balance 0, got %d", penalties[i])
+		}
+	}
+
+	nonAttestedIndices := []uint64{12, 23, 45, 79}
+	for _, i := range nonAttestedIndices {
+		base, err := baseReward(state, i)
+		if err != nil {
+			t.Errorf("Could not get base reward: %v", err)
+		}
+		wanted := 3 * base
+		// Since all these validators did not attest, they shouldn't get rewarded.
+		if rewards[i] != 0 {
+			t.Errorf("Wanted reward balance 0, got %d", rewards[i])
+		}
+		// Base penalties for not attesting.
+		if penalties[i] != wanted {
+			t.Errorf("Wanted penalty balance %d, got %d", wanted, penalties[i])
+		}
+	}
+}
+
+func TestAttestationDelta_SomeAttestedFinalityDelay(t *testing.T) {
+	helpers.ClearAllCaches()
+	e := params.BeaconConfig().SlotsPerEpoch
+	validatorCount := params.BeaconConfig().DepositsForChainStart / 8
+	state := buildState(e+4, validatorCount)
+	startShard := uint64(960)
+	atts := make([]*pb.PendingAttestation, 3)
+	for i := 0; i < len(atts); i++ {
+		atts[i] = &pb.PendingAttestation{
+			Data: &pb.AttestationData{
+				Crosslink: &pb.Crosslink{
+					Shard:    startShard + uint64(i),
+					DataRoot: []byte{'A'},
+				},
+				Target: &pb.Checkpoint{},
+				Source: &pb.Checkpoint{},
+			},
+			AggregationBits: []byte{0xC0, 0xC0, 0xC0, 0xC0},
+			InclusionDelay:  1,
+		}
+	}
+	state.PreviousEpochAttestations = atts
+	state.FinalizedCheckpoint.Epoch = 0
+	state.CurrentCrosslinks[startShard] = &pb.Crosslink{
+		DataRoot: []byte{'A'},
+	}
+	state.CurrentCrosslinks[startShard+1] = &pb.Crosslink{
+		DataRoot: []byte{'A'},
+	}
+
+	rewards, penalties, err := attestationDelta(state)
+	if err != nil {
+		t.Fatal(err)
+	}
 
 	attestedBalance, err := AttestingBalance(state, atts)
 	totalBalance, _ := helpers.TotalActiveBalance(state)
```

Comment:
Can you fix this while you are here please? Handle error

Reply:
Went through the whole file and made sure all errors were covered.